app/controllers/incomeSources/add/AddIncomeSourceStartDateCheckController.scala

```diff
@@ -0,0 +1,288 @@
+      Future(
+        Ok(
+          customNotFoundErrorView()
+        )
+      )
+    } else {
+      (getStartDate(incomeSourceType), getCalls(isAgent, incomeSourceType)) match {
+        case (None, _) =>
+          Logger("application").error(s"[AddIncomeSourceStartDateCheckController][handleSubmitRequest]: " +
+            s"failed to get start date from session for incomeSourceType: $incomeSourceType")
+          Future(
+            getErrorHandler(isAgent).showInternalServerError()
+          )
+        case (Some(startDate), (backCall, postAction, successCall)) =>
+
+          form(incomeSourceType.addIncomeSourceStartDateCheckMessagesPrefix)
+            .bindFromRequest().fold(
+              formWithErrors =>
+                Future(
+                  BadRequest(
+                    addIncomeSourceStartDateCheckView(
+                      form = formWithErrors,
+                      postAction = postAction,
+                      backUrl = backCall.url,
+                      isAgent = isAgent,
+                      incomeSourceStartDate = longDate(startDate.toLocalDate).toLongDate
+                    )
+                  )
+                ),
+              _.toFormMap(form.response).headOption match {
+                case Some(form.responseNo) =>
+                  Future.successful(
+                    Redirect(backCall)
+                      .removingFromSession(
+                        incomeSourceType match {
+                          case SelfEmployment => businessStartDate
+                          case UkProperty => addUkPropertyStartDate
+                          case ForeignProperty => foreignPropertyStartDate
+                        }
+                      )
+                  )
+                case Some(form.responseYes) if incomeSourceType == SelfEmployment =>
+
+                  val businessAccountingPeriodEndDate = dateService
+                    .getAccountingPeriodEndDate(
+                      LocalDate.parse(startDate)
+                    )
+
+                  Future.successful(
+                    Redirect(successCall)
+                      .addingToSession(
+                        SessionKeys.addBusinessAccountingPeriodStartDate -> startDate,
+                        SessionKeys.addBusinessAccountingPeriodEndDate -> businessAccountingPeriodEndDate
+                      )
+                  )
+                case Some(form.responseYes) =>
+                  Future.successful(
+                    Redirect(successCall)
+                  )
+              }
+            )
+      }
+    }
+  } recover {
+    case ex: Exception =>
+      Logger("application").error(s"[AddIncomeSourceStartDateCheckController][handleSubmitRequest]: " +
+        s"Error getting AddIncomeSourceStartDateCheck page: ${ex.getMessage}")
+      getErrorHandler(isAgent).showInternalServerError()
+  }
+
+  private def getStartDate(incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_]): Option[String] = {
```

Comment:
few things can be done here:
* you can wrap results into Future as this would be extracted from the mongo (this changes is optional);
* attempt to convert String into LocalDate, as you are using LocalDate down in the context; so result can be something like String => Either[Throwable, LocalDate]

Reply:
we are planning work related to mongo in the future, so we would read user stored session data from mongo in the future; in order to support this transition in we would need these method to return Future, but this is really something optional at the moment;